### PR TITLE
Add derivation path option on account creation/import.

### PR DESCRIPTION
**Pull request by vanruch**

### CODE REVIEW

File: packages/extension-ui/src/Popup/CreateAccount/AccountCredentials.tsx [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  const [pairType, setPairType] = useState<KeypairType | undefined>(undefined);
+
+  const ref = useRef<HTMLDivElement | null>(null);
+
+  const scroll = (way: 'up' | 'down'): void => {
+    return ref?.current?.scrollTo({
+      top: way === 'up' ? -10000000 : 100000000, // a sufficiently large number
+      behavior: 'smooth'
+    });
+  };
```
Comment (jacogr) [inline packages/extension-ui/src/Popup/CreateAccount/AccountCredentials.tsx]:
```suggestion
  const scroll = useCallback((way: 'up' | 'down'): void => {
    return ref?.current?.scrollTo({
      top: way === 'up' ? -10000000 : 100000000, // a sufficiently large number
      behavior: 'smooth'
    });
  }, []);
```

I know you are just following in the path I set out, but this is actually something we need to start doing. So at least when we start new components (or touch old), let's start doing thing "the-right-way(tm)" :)

File: packages/extension-ui/src/Popup/CreateAccount/AccountCredentials.tsx [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  function _onChange (suriResult: OnDerivationPathChangeProps): void {
+    setSuri(suriResult.suri);
+    if (suriResult.isValid) {
+      setAddress(suriResult.address);
+      setPairType(suriResult.keyPairType);
+    } else {
+      setAddress(null);
+    }
+  }
```
Comment (jacogr) [inline packages/extension-ui/src/Popup/CreateAccount/AccountCredentials.tsx]:
```suggestion
  const _onChange = useCallback((suriResult: OnDerivationPathChangeProps): void {
    setSuri(suriResult.suri);
    if (suriResult.isValid) {
      setAddress(suriResult.address);
      setPairType(suriResult.keyPairType);
    } else {
      setAddress(null);
    }
  }, []);
```

File: packages/extension-ui/src/Popup/CreateAccount/index.tsx [hunk 3]
```diff
@@ -51,11 +51,13 @@ export default function CreateAccount (): React.ReactElement {
...
-        <Mnemonic seed={account.seed} onNextStep={_onNextStep} />
-      ) : (
-        <AccountName address={account.address} onCreate={_onCreate} />
-      ))}</Loading>
+      <Loading>
+        {account && (step === 1 ? (
+          <Mnemonic seed={account.seed} onNextStep={_onNextStep} />
+        ) : (
+          <AccountCredentials seed={account.seed} onCreate={_onCreate} />
+        ))}
```
Comment (jacogr) [inline packages/extension-ui/src/Popup/CreateAccount/index.tsx]:
```suggestion
        {account && (step === 1)
          ? <Mnemonic seed={account.seed} onNextStep={_onNextStep} />
          : <AccountCredentials seed={account.seed} onCreate={_onCreate} />
        }
```

Just small refactoring, easier to read (imho) and fits in what we have elsewhere in this org.

File: packages/extension-ui/src/Popup/ImportSeed.test.tsx [hunk 4]
```diff
@@ -0,0 +1,129 @@
+import React from 'react';
```
Comment (jacogr) [inline packages/extension-ui/src/Popup/ImportSeed.test.tsx]: Missing copyright header.

File: packages/extension-ui/src/Popup/ImportSeed.test.tsx [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+import Adapter from 'enzyme-adapter-react-16';
+
+import { configure, mount, ReactWrapper } from 'enzyme';
```
Comment (jacogr) [inline packages/extension-ui/src/Popup/ImportSeed.test.tsx]:
```suggestion
import { configure, mount, ReactWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
```

File: packages/extension-ui/src/Popup/ImportSeed.test.tsx [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+  Input,
+  Button
+} from '@polkadot/extension-ui/components';
+import { ThemeProvider } from 'styled-components';
+import Import from './ImportSeed';
+import { flushAllPromises } from '../testHelpers';
+import { act } from 'react-dom/test-utils';
+import { BrowserRouter } from 'react-router-dom';
+import DerivationPath, { OptionsLabel } from '@polkadot/extension-ui/partials/DerivationPath';
+import * as messaging from '@polkadot/extension-ui/messaging';
```
Comment (jacogr) [inline packages/extension-ui/src/Popup/ImportSeed.test.tsx]:
```suggestion
import { act } from 'react-dom/test-utils';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';

import {
  ActionContext,
  themes,
  InputWithLabel,
  Input,
  Button
} from '@polkadot/extension-ui/components';
import DerivationPath, { OptionsLabel } from '@polkadot/extension-ui/partials/DerivationPath';
import * as messaging from '@polkadot/extension-ui/messaging';

import Import from './ImportSeed';
import { flushAllPromises } from '../testHelpers';
```
Small import reformat.

File: packages/extension-ui/src/Popup/ImportSeed.tsx [hunk 7]
```diff
@@ -2,39 +2,63 @@
...
-  TextAreaWithLabel,
-  VerticalSpace
+  TextAreaWithLabel
 } from '../components';
 import { createAccountSuri, validateSeed } from '../messaging';
-import { Name, Password } from '../partials';
+import { Name, Password, DerivationPath } from '../partials';
 import styled from 'styled-components';
+import { OnDerivationPathChangeProps } from '@polkadot/extension-ui/partials/DerivationPath';
+import { KeypairType } from '@polkadot/util-crypto/types';
```
Comment (jacogr) [inline packages/extension-ui/src/Popup/ImportSeed.tsx]: Move these to below the react import. Always -

- import from names packages
- <blank line>
- import from relative paths

(alphabetical in these sections)

File: packages/extension-ui/src/Popup/ImportSeed.tsx [hunk 8]
```diff
@@ -2,39 +2,63 @@
...
+  const [address, setAddress] = useState<string | null>(null);
+  const [suri, setSuri] = useState<string | null>(null);
+  const [pairType, setPairType] = useState<KeypairType | undefined>(undefined);
+  const ref = useRef<HTMLDivElement | null>(null);
 
-  const _onChangeSeed = (suri: string): Promise<void> =>
-    validateSeed(suri)
-      .then(setAccount)
-      .catch((): void => setAccount(null));
+  const scroll = (way: 'up' | 'down'): void => {
```
Comment (jacogr) [inline packages/extension-ui/src/Popup/ImportSeed.tsx]: As per the useCallback comments, for this and the next.

File: packages/extension-ui/src/components/InputWithLabel.tsx [hunk 9]
```diff
@@ -19,9 +19,10 @@ interface Props {
   onChange?: (value: string) => void;
   type?: 'text' | 'password';
   value?: string;
+  placeholder?: string;
 }
 
-function InputWithLabel ({ className, defaultValue, label, isFocused, isReadOnly, onBlur, onChange, type = 'text', value, isError }: Props): React.ReactElement<Props> {
+function InputWithLabel ({ className, defaultValue, label, isFocused, isReadOnly, onBlur, onChange, type = 'text', value, isError, placeholder }: Props): React.ReactElement<Props> {
```
Comment (jacogr) [inline packages/extension-ui/src/components/InputWithLabel.tsx]:
```suggestion
function InputWithLabel ({ className, defaultValue, label, isError, isFocused, isReadOnly, onBlur, onChange, placeholder, type = 'text', value }: Props): React.ReactElement<Props> {
```

Alphabetical.

File: packages/extension-ui/src/partials/DerivationPath.tsx [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+// This software may be modified and distributed under the terms
+// of the Apache-2.0 license. See the LICENSE file for details.
+
+import React, { useState, useEffect } from 'react';
+import gearIcon from '../assets/gear.svg';
+import arrowIcon from '../assets/arrow-down.svg';
+import styled from 'styled-components';
+import { KeypairType } from '@polkadot/util-crypto/types';
+import { Dropdown, InputWithLabel, Svg } from '@polkadot/extension-ui/components';
+import { validateSeed } from '@polkadot/extension-ui/messaging';
```
Comment (jacogr) [inline packages/extension-ui/src/partials/DerivationPath.tsx]:
```suggestion
import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import { KeypairType } from '@polkadot/util-crypto/types';
import { Dropdown, InputWithLabel, Svg } from '@polkadot/extension-ui/components';
import { validateSeed } from '@polkadot/extension-ui/messaging';

import arrowIcon from '../assets/arrow-down.svg';
import gearIcon from '../assets/gear.svg';
```

File: packages/extension-ui/src/partials/DerivationPath.tsx [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+  }
+};
+
+function DerivationPath ({ onChange, onDoScroll, seed }: Props): React.ReactElement<Props> {
+  const [type, setType] = useState<KeypairType | undefined>(undefined);
+  const [path, setPath] = useState('');
+  const [isValid, setValid] = useState(false);
+  const [showAdvancedOptions, setShowAdvancedOptions] = useState(false);
+
+  const getSuri = (): string => (path ? `${seed}${path}` : seed);
```
Comment (jacogr) [inline packages/extension-ui/src/partials/DerivationPath.tsx]: useCallback (I know, I know - all my fault :))

File: packages/extension-ui/src/partials/DerivationPath.tsx [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+
+  useEffect(() => {
+    const suri = getSuri();
+    (async (): Promise<void> => {
+      try {
+        const { address } = await validateSeed(suri, type);
+        onChange({
```
Comment (jacogr) [inline packages/extension-ui/src/partials/DerivationPath.tsx]:
```suggestion
        const { address } = await validateSeed(suri, type);
        
        onChange({
```

**Comment by jacogr** [comment]
My functional feedback -

- as it stands the derivation is completely hidden on creation, I knew it was there but took me 2 tries to actually locate it since I needed to scroll
- we need to stick to sr25519 here only - it is actually not an option that atm we need to provide. Really trying to keep it as simple as possible.

Overall, the better approach imho (and something we need to do anyway) is to rather, allow derivation from a specific account. So i.e. in our account options (forget/export), we can select an account and derive another from it. So the benefits we get -

- this puts us down the road of having to manage a single root account only
- when deriving sub-accounts, you don't need to repeat anything
- here is where sr25519 is nice - since it has both hard/soft, while ed only has a single type

**Comment by vanruch** [comment]
Closing this one, please see https://github.com/polkadot-js/extension/pull/296 instead

**Pull request closed without merging**